File: nodejs/ai-convert.js
// Experimental: convert a plain PDF invoice to Factur-X using AI.
// Review output manually before sending to a customer or tax authority.
//
// Get API key: https://www.invoicexml.com/account/authentication
// Docs:        https://www.invoicexml.com/docs/api/transform/to/facturx
//
// Requires Node.js 18+. No dependencies.

const fs = require('fs');

const apiKey = 'YOUR_API_KEY';

(async () => {
    const form = new FormData();
    form.append('file', new Blob([fs.readFileSync('plain-invoice.pdf')], { type: 'application/pdf' }), 'plain-invoice.pdf');
    form.append('version',  '2.3.2');
    form.append('profile',  'extended');
    form.append('syntax',   'cii');
    form.append('embed',    'true');
    form.append('language', 'en');

    const response = await fetch('https://api.invoicexml.com/v1/transform/to/facturx', {
        method: 'POST',
        headers: { Authorization: `Bearer ${apiKey}` },
        body: form,
    });

    if (!response.ok) {
        console.error(`InvoiceXML API error ${response.status}: ${await response.text()}`);
        process.exit(1);
    }

    const buffer = Buffer.from(await response.arrayBuffer());
    fs.writeFileSync('converted-facturx.pdf', buffer);
    console.log(`Saved converted-facturx.pdf (${buffer.length} bytes)`);
    console.log('Review the output manually before sending.');
})();

File: csharp/AiConvert.cs
// EXPERIMENTAL: Convert a plain (unstructured) PDF invoice into a fully
// compliant Factur-X PDF/A-3 with embedded XML, using the InvoiceXML
// AI-powered transform API.
//
// WARNING: Real-world PDF invoices are often messy: scanned, irregularly
// formatted, multi-page, multilingual, or missing fields that EN 16931
// requires. AI extraction can make subtle mistakes (wrong tax codes,
// transposed amounts, missing VAT IDs, currency issues) that automated
// validators may not catch. ALWAYS review the output PDF before sending
// it to a customer or tax authority.
//
// Get your API key at: https://www.invoicexml.com/account/authentication
// Full API reference:  https://www.invoicexml.com/docs/api/transform/to/facturx
//
// Required NuGet package:
//   dotnet add package Flurl.Http


using Flurl.Http;

public static class AiConvert
{
    private const string Endpoint = "https://api.invoicexml.com/v1/transform/to/facturx";

    public static async Task<byte[]> RunAsync(
        string inputPdfPath = "plain-invoice.pdf",
        string outputPath = "converted-facturx.pdf",
        string version = "2.3.2",
        string profile = "extended",
        string syntax = "cii",
        string language = "en")
    {
        // Get an InvoiceXML API key at: https://www.invoicexml.com/account/authentication
        var apiKey = "INVOICEXML_API_KEY";

        if (!File.Exists(inputPdfPath))
            throw new FileNotFoundException($"Input PDF not found: {inputPdfPath}");

        try
        {
            // POST the plain PDF; InvoiceXML extracts invoice fields with AI,
            // builds an EN 16931 XML payload, validates it, and embeds it
            // into a PDF/A-3 container. Parameter reference:
            // https://www.invoicexml.com/docs/api/transform/to/facturx
            var pdfBytes = await Endpoint
                .WithOAuthBearerToken(apiKey)
                .PostMultipartAsync(mp => mp
                    .AddFile("file", inputPdfPath, "application/pdf")
                    .AddString("version",  version)   // e.g. "2.3.2"
                    .AddString("profile",  profile)   // minimum, basicwl, basic, en16931, extended
                    .AddString("syntax",   syntax)    // "cii" for Factur-X / ZUGFeRD
                    .AddString("embed",    "true")    // embed XML into the returned PDF
                    .AddString("language", language)  // hint for AI extraction
                )
                .ReceiveBytes();

            await File.WriteAllBytesAsync(outputPath, pdfBytes);
            Console.WriteLine($"Converted Factur-X saved: {outputPath} ({pdfBytes.Length:N0} bytes)");
            Console.WriteLine("REMINDER: review the output manually before sending to a customer or tax authority.");
            return pdfBytes;
        }
        catch (FlurlHttpException ex)
        {
            var body = await ex.GetResponseStringAsync();
            Console.Error.WriteLine($"InvoiceXML API error {(int?)ex.StatusCode}: {body}");
            throw;
        }
    }
}
